shell/platform/windows/platform_handler.cc

```diff
@@ -18,11 +20,123 @@ static constexpr char kSetClipboardDataMethod[] = "Clipboard.setData";
 static constexpr char kTextPlainFormat[] = "text/plain";
 static constexpr char kTextKey[] = "text";
 
-static constexpr char kUnknownClipboardFormatError[] =
-    "Unknown clipboard format error";
+static constexpr char kClipboardError[] = "Clipboard error";
+static constexpr char kUnknownClipboardFormatMessage[] =
+    "Unknown clipboard format";
 
 namespace flutter {
 
+namespace {
+
+// A scoped wrapper for GlobalLock/GlobalUnlock.
+class ScopedGlobalLock {
+ public:
+  // Attempts to acquire a global lock on |memory| for the life of this object.
+  ScopedGlobalLock(HGLOBAL memory) {
+    source_ = memory;
+    locked_memory_ = ::GlobalLock(memory);
+  }
+
+  ~ScopedGlobalLock() {
+    if (locked_memory_) {
+      ::GlobalUnlock(source_);
+    }
+  }
+
+  // Returns the locked memory pointer, which will be nullptr if acquiring the
+  // lock failed.
+  void* get() { return locked_memory_; }
+
+ private:
+  HGLOBAL source_;
+  void* locked_memory_;
+};
+
+// A Clipboard wrapper that automatically closes the clipboard when it goes out
+// of scope.
+class ScopedClipboard {
+ public:
+  ScopedClipboard();
+  ~ScopedClipboard();
+
+  // Attempts to open the clipboard for the given window, returning true if
+  // successful.
+  bool Open(HWND window);
+
+  // Returns true if there is string data available to get.
+  bool HasString();
+
+  // Returns string data from the clipboard.
+  //
+  // If getting a string fails, returns false and does not change |out_string|.
+  // Get error information with ::GetLastError().
+  //
+  // Open(...) must have succeeded to call this method.
+  bool GetString(std::wstring* out_string);
+
+  // Sets the string content of the clipboard, returning true on success.
+  //
+  // On failure, get error information with ::GetLastError().
+  //
+  // Open(...) must have succeeded to call this method.
+  bool SetString(const std::wstring string);
+
+ private:
+  bool opened_ = false;
+};
+
+ScopedClipboard::ScopedClipboard() {}
+
+ScopedClipboard::~ScopedClipboard() {
+  if (opened_) {
+    ::CloseClipboard();
+  }
+}
+
+bool ScopedClipboard::Open(HWND window) {
+  opened_ = ::OpenClipboard(window);
+  return opened_;
+}
+
+bool ScopedClipboard::HasString() {
+  // Allow either plain text format, since getting data will auto-interpolate.
+  return ::IsClipboardFormatAvailable(CF_UNICODETEXT) ||
+         ::IsClipboardFormatAvailable(CF_TEXT);
+}
+
+bool ScopedClipboard::GetString(std::wstring* out_string) {
+  assert(out_string != nullptr);
+  assert(opened_);
```

Comment:
Is not being able to open a clipboard such an exceptional event that you'd take down the process? Maybe just return false?

Reply:
> It's a programming error not to open it first

I did not realize this.